File: LeetCode/iter1/c++/241.different-ways-to-add-parentheses.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <vector>
#include <sstream>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <queue>
#include <stack>

using namespace std;

class Solution {
public:
	vector<int> diffWaysToCompute(string input) {
		return diffWaysToComputeHelper(input, 0, input.size() - 1);
	}
private:
	vector<int> diffWaysToComputeHelper(string s, int startIdx, int endIdx) {
		vector<int> v;
		int num = isAllNumber(s, startIdx, endIdx);
		if (num != -1) {
			v.push_back(num);
			return v;
		}

		for (int i = startIdx; i <= endIdx; i++) {
			if (s[i] == '+' || s[i] == '-' || s[i] == '*') {
				vector<int> leftV = diffWaysToComputeHelper(s, startIdx, i - 1);
				vector<int> rightV = diffWaysToComputeHelper(s, i + 1, endIdx);

				for (int leftVal : leftV) {
					for (int rightVal : rightV) {
						if (s[i] == '+') v.push_back(leftVal + rightVal);
						if (s[i] == '-') v.push_back(leftVal - rightVal);
						if (s[i] == '*') v.push_back(leftVal * rightVal);
					}
				}
			}
		}

		return v;
	}


	int isAllNumber(string input, int startIdx, int endIdx) {
		int ret = 0, digit = 1;
		for (int i = endIdx; i >= startIdx; i--) {
			if (input[i] < '0' || input[i] > '9') {
				ret = -1;
				break;
			}

			ret += (input[i] - '0') * digit;
			digit *= 10;
		}

		return ret;
	}
};

void testPrint(vector<int> v) {
	for (int num : v) {
		cout << num << " ";
	}
	cout << endl;
}

int main(void)
{
	Solution s;
	testPrint(s.diffWaysToCompute("2-1-1"));
	testPrint(s.diffWaysToCompute("2*3-4*5"));

	return 0;
}

File: LeetCode/iter1/java/241.different-ways-to-add-parentheses.java
package org.ironman.base.playground.leetcode;

import java.util.ArrayList;
import java.util.List;

public class Solution {

    public static void main(String[] args) {
        System.out.println(diffWaysToCompute("2-1-1"));
        System.out.println(diffWaysToCompute("2*3-4*5"));
    }

    public static List<Integer> diffWaysToCompute(String input) {
        List<Integer> ret = new ArrayList<>();
        int num = str2int(input);
        if (num != -1) {
            ret.add(num);
            return ret;
        }

        for (int i = 0; i < input.length(); i++) {
            if (input.charAt(i) == '+' || input.charAt(i) == '-' || input.charAt(i) == '*') {
                List<Integer> leftLst = diffWaysToCompute(input.substring(0, i));
                List<Integer> rightLst = diffWaysToCompute(input.substring(i + 1));
                for (int leftVal : leftLst) {
                    for (int rightVal : rightLst) {
                        if (input.charAt(i) == '+') {
                            ret.add(leftVal + rightVal);
                        } else if (input.charAt(i) == '-') {
                            ret.add(leftVal - rightVal);
                        } else if (input.charAt(i) == '*') {
                            ret.add(leftVal * rightVal);
                        }
                    }
                }
            }
        }

        return ret;
    }

    private static int str2int(String input) {
        int ret = 0, digit = 1;
        for (int i = input.length() - 1; i >= 0; i--) {
            if (input.charAt(i) < '0' || input.charAt(i) > '9') {
                ret = -1;
                break;
            }

            ret += digit * (input.charAt(i) - '0');
            digit *= 10;
        }

        return ret;
    }
}